### DISCUSSION TITLE
Applying the same animation to objects in a collection through a loop

### DISCUSSION THREAD

**Discussion by SushkoTech** [KyryloKuzyk/PrimeTween#104] [FAQ] [answered]
Hey! I've been using your Asset for some time now, it really came in handy!  I recently run into a problem when trying to loop through a list of objects in coroutine, in order to apply the same animation to all these objects except the last one.

```c#
private IEnumerator DestroyObjects()
{
    for (int i = 0; i < _objects.Count - 1; i++)
    {
        yield return Tween.Scale(
        _objects[i].transform, 
        startValue: 1f, 
        endValue: 0f, 
        duration: 0.2f)
       .OnComplete(() => Destroy(_objects[i]))
       .ToYieldInstruction();
    }
}
```
The problem is that the animation is only applied to the first object in the list. All the others are simply destroyed when the specified `duration` in `Tween` is reached. How can I loop through objects in a collection and apply the same animation to them?

**Comment by KyryloKuzyk** [comment, marked as answer]
Hey! I'm glad PrimeTween is helpful for you :)

Hmm, I just copy-pasted your code and it works as expected: all objects except the last one are scaled down, then destroyed one by one.

However, I was able to replicate a similar issue to what you're describing by removing the `yield return ToYieldInstruction()` part. In this case, all objects are scaled down, but only the last one is destroyed. This happens because the `i` variable is captured in a closure inside the OnComplete() callback and its value is shared between all callbacks. The compiler also warns about this issue with the warning: "Captured variable is modified in the outer scope".

There are two ways how you can fix the issue. The first and dirty one is to copy the `i` variable into some new variable called `indexCopy` and use that copied variable inside the OnComplete().

But a cleaned solution would be to use OnComplete() overload that accepts target as the first parameter, like so:
```csharp
for (int i = 0; i < _objects.Count - 1; i++) {
    Tween.Scale(
            _objects[i].transform,
            startValue: 1f,
            endValue: 0f,
            duration: 0.2f)
        .OnComplete(_objects[i], target => Destroy(target));
}
```

But still, your code seems to work fine, and because all the animations play sequentially due to `ToYieldInstruction()` call, the captured variable doesn't make a difference. I would still fix the compiler warning though.

**Reply by SushkoTech** [reply]
Thank you for fast response! I will check all of this tomorrow and will let you know :)

**Reply by SushkoTech** [reply]
I checked it all out again. And the problem was, that on my prefab that used all the objects in the list, had the Unity animator attached. I forgot to remove it after the old development stages. Now everything works perfectly! Thanks for the fast response and for suggesting a cleaner solution for `OnComplete()`:  `OnComplete(_objects[i], target => Destroy(target))`. Sorry for wasting your time on this.
Wishing PrimeTween to become the main standard for developers in the future! And also hi from 🇺🇦 to 🇺🇦 i guess :)
